### PR TITLE
Cleanup dependency check; fixes #977

### LINKED ISSUE #977: Cleanup dependency check
Instead of a check for each needed extension, cleanup this by a check in a loop

Proposal:

```
$needed_extensions = [
    [ 'name' => 'session' ], 
    [ 'name' => 'ctype', 'func' => 'ctype_digit' ], 
    [ 'name' => 'fileinfo' , 'class' => 'finfo' ],
    [ 'name' => 'zlib' ], 
    [ 'name' => 'gd' ], 
    [ 'name' => 'mbstring' ], 
    [ 'name' => 'json' , 'func' => 'json_encode' ],
    ...
}
foreach ($needed_extensions as ext) {
   if (isset($ext['func']) ... function_exists
   else if (isset($ext['class']) ... class_exists
   else ... extension_loaded
}

```

### PR DESCRIPTION
**Pull request by trasher**

### CODE REVIEW

File: inc/toolbox.class.php [hunk 1]
```diff
@@ -941,111 +927,87 @@ static function commonCheckForUseGLPI() {
...
-
-      if (!extension_loaded('curl')) {
-         echo "<td><img src='".$CFG_GLPI['root_doc']."/pics/ko_min.png'>".
-                     __('Curl extension of your parser PHP is not installed')."></td>";
-         $error = 2;
+      $needed_extentions = [
+         'mysqli'   => [
+            'title'     => __('MySQL Improved extension test'),
+            'success'   => __s('Ok - the MySQLi class exist - Perfect!'),
+            'error'     => __('You must install the MySQL Improved extension for PHP.')
```
Comment (remicollet) [inline inc/toolbox.class.php]: I think this should be 1 or 2 (the error return)
Then same message can be used for all

Comment (remicollet) [inline inc/toolbox.class.php]: I mean 1=optional, 2=mandatory (see install.php)

File: inc/toolbox.class.php [hunk 2]
```diff
@@ -941,111 +927,87 @@ static function commonCheckForUseGLPI() {
...
-                     __s('The functionality is found - Perfect!'). "\" title=\"".
-                     __s('The functionality is found - Perfect!')."\"></td>";
+         echo "<tr class=\"tab_bg_1\"><td class=\"left b\">{$params['title']}</td>";
+         if ($success === true) {
+            echo "<td><img src=\"{$CFG_GLPI['root_doc']}/pics/ok_min.png\"
+                    alt=\"{$params['success']}\"
+                    title=\"{$params['success']}\"></td>";
+         } else {
+            echo "<td class=\"red\"><img src=\"{$CFG_GLPI['root_doc']}/pics/ko_min.png\">{$params['error']}</td>";
+            $error = 2;
```
Comment (remicollet) [inline inc/toolbox.class.php]: See previous.

if ($error=1) ....'you should install '
if ($error=2) ....'you MUST install '

File: inc/toolbox.class.php [hunk 3]
```diff
@@ -941,111 +927,87 @@ static function commonCheckForUseGLPI() {
...
+            'title'     => __('Zlib extension test'),
+            'success'   => __s('The functionality is found - Perfect!'),
+            'error'     => __('Zlib extension of your parser PHP is not installed')
+         ],
+         'curl'      => [
+            'title'     => __('Curl extension test'),
+            'success'   => __s('The functionality is found - Perfect!'),
+            'error'     => __('Curl extension of your parser PHP is not installed')
+         ]
+      ];
```
Comment (remicollet) [inline inc/toolbox.class.php]: Il est donc possible d'ajouter imap et ldap qui sont optionnelles (comme gd)

File: inc/toolbox.class.php (hunk 1 above)
Comment (trasher) [inline inc/toolbox.class.php]: Indeed, we got a `$error = 1;` for the `gd` extension test; I did not see it. thank you!

File: inc/toolbox.class.php [hunk 4]
```diff
@@ -941,111 +927,87 @@ static function commonCheckForUseGLPI() {
...
+
+      $loaded_extensions = [];
+      //check for PHP extensions
+      foreach ($needed_extentions as $ext => $params) {
+         $success = true;
+
+         if (isset($params['functions'])) {
+            foreach ($params['functions'] as $function) {
```
Comment (remicollet) [inline inc/toolbox.class.php]: I'm not sure testing various functions is really useful, a single one should be enough.

**Comment by trasher** [comment]
So... Should we merge this one for 9.1; or wait for the next release?

File: inc/toolbox.class.php [hunk 5]
```diff
@@ -941,111 +927,77 @@ static function commonCheckForUseGLPI() {
...
+            }
+         }
 
-      } else {
-         echo "<td><img src='".$CFG_GLPI['root_doc']."/pics/ok_min.png' alt=\"".
-                     __s('The functionality is found - Perfect!'). "\" title=\"".
-                     __s('The functionality is found - Perfect!')."\"></td>";
+         echo "<tr class=\"tab_bg_1\"><td class=\"left b\">{sprintf('%s extension test', $ext)}</td>";
+         if ($success === true) {
+            $msg = sprintf('Extension %s is installed', $ext);
```
Comment (remicollet) [inline inc/toolbox.class.php]: gettext is missing

Comment (trasher) [inline inc/toolbox.class.php]: Hey... Well catched, thanks :)

File: inc/toolbox.class.php [hunk 6]
```diff
@@ -941,111 +927,77 @@ static function commonCheckForUseGLPI() {
...
+         echo "<tr class=\"tab_bg_1\"><td class=\"left b\">{sprintf('%s extension test', $ext)}</td>";
+         if ($success === true) {
+             $msg = sprintf(__('Extension %s is installed'), $ext);
+            echo "<td><img src=\"{$CFG_GLPI['root_doc']}/pics/ok_min.png\"
+                    alt=\"$msg\"
+                    title=\"$msg\"></td>";
+         } else {
+            if (isset($params['required']) && $params['required'] === true) {
+               $error = 2;
+               echo "<td class=\"red\"><img src=\"{$CFG_GLPI['root_doc']}/pics/ko_min.png\">{sprintf('Extension %s is missing', $ext)}</td>";
```
Comment (remicollet) [inline inc/toolbox.class.php]: Also missig here ;)

File: inc/toolbox.class.php [hunk 7]
```diff
@@ -941,111 +927,77 @@ static function commonCheckForUseGLPI() {
...
+            } else {
+               $error = 1;
+               echo "<td><img src=\"{$CFG_GLPI['root_doc']}/pics/warning_min.png\">{sprintf('Extension %s is not present', $ext)}</td>";
```
Comment (remicollet) [inline inc/toolbox.class.php]: and here :p

File: inc/toolbox.class.php [hunk 8]
```diff
@@ -941,111 +927,77 @@ static function commonCheckForUseGLPI() {
...
+            if (extension_loaded($ext)) {
+               $success = false;
+            }
+         }
 
-      } else {
-         echo "<td><img src='".$CFG_GLPI['root_doc']."/pics/ok_min.png' alt=\"".
-                     __s('The functionality is found - Perfect!'). "\" title=\"".
-                     __s('The functionality is found - Perfect!')."\"></td>";
+         echo "<tr class=\"tab_bg_1\"><td class=\"left b\">{sprintf(__('%s extension test'), $ext)}</td>";
```
Comment (remicollet) [inline inc/toolbox.class.php]: doesn't work, I think you cannot escape function call in this way, only var.

Comment (trasher) [inline inc/toolbox.class.php]: I've fixed that, thanks :)

**Comment by trasher** [comment]
OK, ths is not corect. `$error` can be overriden... Will have to rework.

**Comment by trasher** [comment]
OK, I've reworked that... I had inverted a true and false (so, yes, that was not working as expected :D).

I've also changed string to "%s extension" for all.

**Comment by remicollet** [comment]
Thanks for your patience ;)

Merged (I just add a space between the icon and the message)

**Comment by trasher** [comment]
Thanks for your patience :D

**Pull request closed without merging**